### PR TITLE
Fix remaining memory leaks in C++ tests

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Fixes the remaining memory leaks discovered by address sanitizer when running the C++ unit tests.

**To test:**

Code review.
See #27506 for details of the required suppression file for running the tests

*There is no associated issue.*

*This does not require release notes*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by DavidFair — CHANGES_REQUESTED** [review]
Some minor additional clean-up bits I've noticed

File: Framework/Algorithms/test/DetectorEfficiencyCorTest.h [hunk 1]
```diff
@@ -134,14 +134,15 @@ class DetectorEfficiencyCorTest : public CxxTest::TestSuite {
     const int ndets(2);
     std::vector<Detector *> detectors;
     for (int i = 0; i < ndets; ++i) {
-      Detector *detector = new Detector("det", i + 1, shape, nullptr);
+      Detector *detector = new Detector("det", i + 1, shape, instrument.get());
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Algorithms/test/DetectorEfficiencyCorTest.h]: I agree here but I was a little concerned in touching the whole tree in this PR and I don't have time to do that now. I'd be happy for that to be a maintenance task for the future though.

File: Framework/Algorithms/src/FilterEvents.cpp [hunk 2]
```diff
@@ -788,10 +789,13 @@ void FilterEvents::splitTimeSeriesProperty(
       g_log.information() << "Workspace target (" << tindex
                           << ") does not have workspace associated."
                           << "\n";
+      delete output_vector[tindex];
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Algorithms/src/FilterEvents.cpp]: Thanks for catching this. I don't know why I didn't change it as I changed `split_tsp_vector` to use `unique_ptr`.

[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/Algorithms/src/FilterEvents.cpp]: Now I see why I did this. `tsp->splitByTimeVector` on line 778 requires a vector of raw pointers. I can construct a temporary of those as it seems better for exception handling to have the `unique_ptr` handle the memory.

**Review by DavidFair — APPROVED** [review]

**Pull request merged**